src/database-sqlite3.cpp

```diff
@@ -62,88 +68,484 @@ SQLite format specification:
+			<< cur_time - bhd->first_time << " ms." << std::endl;
+	} else if (cur_time - bhd->first_time >= BUSY_WARNING_TRESHOLD &&
+			bhd->prev_time - bhd->first_time < BUSY_WARNING_TRESHOLD) {
 		warningstream << "SQLite3 database has been locked for "
-			<< cur_time - first_time << " ms." << std::endl;
-	} else if (cur_time - first_time >= BUSY_ERROR_TRESHOLD &&
-			prev_time - first_time < BUSY_ERROR_TRESHOLD) {
+			<< cur_time - bhd->first_time << " ms." << std::endl;
+	} else if (cur_time - bhd->first_time >= BUSY_ERROR_TRESHOLD &&
+			bhd->prev_time - bhd->first_time < BUSY_ERROR_TRESHOLD) {
 		errorstream << "SQLite3 database has been locked for "
-			<< cur_time - first_time << " ms; this causes lag." << std::endl;
-	} else if (cur_time - first_time >= BUSY_FATAL_TRESHOLD &&
-			prev_time - first_time < BUSY_FATAL_TRESHOLD) {
+			<< cur_time - bhd->first_time << " ms; this causes lag." << std::endl;
+	} else if (cur_time - bhd->first_time >= BUSY_FATAL_TRESHOLD &&
+			bhd->prev_time - bhd->first_time < BUSY_FATAL_TRESHOLD) {
 		errorstream << "SQLite3 database has been locked for "
-			<< cur_time - first_time << " ms - giving up!" << std::endl;
-	} else if ((cur_time - first_time) / BUSY_ERROR_INTERVAL !=
-			(prev_time - first_time) / BUSY_ERROR_INTERVAL) {
+			<< (cur_time - bhd->first_time) / 1000 << " seconds - giving up!" << std::endl;
+	} else if ((cur_time - bhd->first_time) / BUSY_ERROR_INTERVAL !=
+			(bhd->prev_time - bhd->first_time) / BUSY_ERROR_INTERVAL) {
 		// Safety net: keep reporting every BUSY_ERROR_INTERVAL
 		errorstream << "SQLite3 database has been locked for "
-			<< (cur_time - first_time) / 1000 << " seconds!" << std::endl;
+			<< (cur_time - bhd->first_time) / 1000 << " seconds!" << std::endl;
 	}
 
-	prev_time = cur_time;
+	bhd->prev_time = cur_time;
 
 	// Make sqlite transaction fail if delay exceeds BUSY_FATAL_TRESHOLD
-	return cur_time - first_time < BUSY_FATAL_TRESHOLD;
+	return cur_time - bhd->first_time < BUSY_FATAL_TRESHOLD;
+}
+
+
+void SQLite3::beginTransaction() {
+	SQLRES(sqlite3_step(m_stmt_begin), SQLITE_DONE,
+		"Failed to start SQLite3 transaction");
+	sqlite3_reset(m_stmt_begin);
+}
+
+
+void SQLite3::commitTransaction() {
+	SQLRES(sqlite3_step(m_stmt_commit), SQLITE_DONE,
+		"Failed to commit SQLite3 transaction");
+	sqlite3_reset(m_stmt_commit);
+}
+
+
+std::string SQLite3::getDBJournalMode()
+{
+	sqlite3_stmt *stmt_journal;
+
+	SQLOK(sqlite3_prepare_v2(m_database, "PRAGMA journal_mode;",
+			-1, &stmt_journal, NULL),
+		"Failed to prepare get-journal-mode query")
+	if (sqlite3_step(stmt_journal) != SQLITE_ROW)
+		throw DatabaseException("SQLite3: Failed to get journal mode from database");
+	std::string mode = (const char *) sqlite3_column_text(stmt_journal, 0);
+	mode = lowercase(mode);
+	SQLOK_ERRSTREAM(sqlite3_finalize(stmt_journal),
+		"SQLite3: Failed to finalize get-journal-mode statement")
+	return mode;
+}
+
+
+void SQLite3::setJournalMode()
```

Comment:
This isn't a very good name since it looks like a setter.  I suggest `applyJournalMode`.